File: apps/mobile/src/domain/eligibility.ts
import type { EligibilityResult, Property } from './types';

export const RULESET_VERSION = 'demeter-triage-0.1.0';

export function assessEligibility(property: Property): EligibilityResult {
  let score = 0;
  const reasons: string[] = [];
  const pending: string[] = [];

  if (['degraded_pasture', 'abandoned', 'agriculture'].includes(property.landUse)) {
    score += 35;
    reasons.push('O uso declarado pode justificar análise de restauração ou transição produtiva.');
  } else if (property.landUse === 'agroforestry') {
    score += 20;
    reasons.push('O sistema agroflorestal existente exige avaliação de linha de base e adicionalidade.');
  } else {
    reasons.push('Vegetação nativa existente não implica remoção adicional automaticamente.');
    pending.push('Avaliar conservação, conformidade e alternativas além de créditos de remoção.');
  }

  if (property.answers.intendsRestoration) {
    score += 30;
    reasons.push('Existe intenção declarada de aumentar ou restaurar cobertura vegetal.');
  } else {
    pending.push('Definir intervenção futura e responsável pela execução.');
  }

  if (property.answers.hasPossessionProof) {
    score += 25;
    reasons.push('Foi declarada existência de documento de vínculo com a área.');
  } else {
    pending.push('Comprovar vínculo, autorização e titularidade dos benefícios ambientais.');
  }

  if (property.answers.recentClearing) {
    score -= 45;
    reasons.push('Supressão recente declarada exige revisão técnica e jurídica antes de qualquer enquadramento.');
    pending.push('Verificar histórico de uso do solo e regularidade da supressão declarada.');
  }

  if (property.points.length < 3) pending.push('Completar perímetro geográfico válido.');
  pending.push('Validar linha de base, adicionalidade, permanência, vazamento e metodologia aplicável.');

  const normalized = Math.max(0, Math.min(100, score));
  if (property.answers.recentClearing || normalized < 30) {
    return { status: 'not_ready', title: 'Ainda não está pronta para agrupamento', summary: 'Há riscos ou lacunas que precisam de revisão antes da análise de projeto.', score: normalized, reasons, pending, rulesetVersion: RULESET_VERSION };
  }
  if (normalized < 70) {
    return { status: 'needs_review', title: 'Potencial condicionado', summary: 'A área pode seguir para revisão, mas faltam evidências e definições importantes.', score: normalized, reasons, pending, rulesetVersion: RULESET_VERSION };
  }
  return { status: 'potential', title: 'Potencial preliminar identificado', summary: 'A área possui sinais iniciais compatíveis com uma avaliação técnica de restauração ou agrofloresta.', score: normalized, reasons, pending, rulesetVersion: RULESET_VERSION };
}

File: services/api/app/domain/eligibility.py
import hashlib
import json

from app.domain.schemas import AssessmentRead, EligibilityStatus, LandUse, AssessmentInput

RULESET_VERSION = "demeter-triage-0.1.0"


def canonical_hash(value: AssessmentInput) -> str:
    payload = json.dumps(value.model_dump(mode="json"), sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def assess(property_data: AssessmentInput) -> AssessmentRead:
    score = 0
    reasons: list[str] = []
    pending: list[str] = []

    if property_data.land_use in {LandUse.DEGRADED_PASTURE, LandUse.ABANDONED, LandUse.AGRICULTURE}:
        score += 35
        reasons.append(
            "O uso declarado pode justificar análise de restauração ou transição produtiva."
        )
    elif property_data.land_use == LandUse.AGROFORESTRY:
        score += 20
        reasons.append("O sistema existente exige avaliação de linha de base e adicionalidade.")
    else:
        reasons.append("Vegetação nativa existente não implica remoção adicional automaticamente.")
        pending.append("Avaliar conservação e alternativas além de créditos de remoção.")

    if property_data.intends_restoration:
        score += 30
        reasons.append("Existe intenção declarada de aumentar ou restaurar cobertura vegetal.")
    else:
        pending.append("Definir intervenção futura e responsável pela execução.")

    if property_data.has_possession_proof:
        score += 25
        reasons.append("Foi declarada existência de documento de vínculo com a área.")
    else:
        pending.append("Comprovar vínculo e titularidade dos benefícios ambientais.")

    if property_data.recent_clearing:
        score -= 45
        reasons.append("Supressão recente exige revisão técnica e jurídica.")
        pending.append("Verificar histórico de uso do solo e regularidade da supressão.")

    pending.append("Validar linha de base, adicionalidade, permanência, vazamento e metodologia.")
    score = max(0, min(100, score))

    if property_data.recent_clearing or score < 30:
        status = EligibilityStatus.NOT_READY
    elif score < 70:
        status = EligibilityStatus.NEEDS_REVIEW
    else:
        status = EligibilityStatus.POTENTIAL

    return AssessmentRead(
        property_id=property_data.property_id,
        status=status,
        score=score,
        reasons=reasons,
        pending=pending,
        ruleset_version=RULESET_VERSION,
        input_hash=canonical_hash(property_data),
    )
